### ISSUE TITLE
High frequency issue with the XieDelhommeau Green function

### ISSUE THREAD

**Issue by mancellin** [capytaine/capytaine#298]
The modification of the "XieDelhommeau" variant of the Green function done in #216 seems to have some issues for high frequencies.

![XieDel_issue](https://user-images.githubusercontent.com/31126826/222381998-8ad1d20d-715f-4c7d-a842-0268f15e612b.png)

<details>
<summary>Source for the above figure</summary>

```python
import numpy as np
import xarray as xr
import capytaine as cpt
import matplotlib.pyplot as plt

mesh = cpt.mesh_vertical_cylinder(center=(0, 0, 0), length=2.0, radius=1.0).immersed_part()
body = cpt.FloatingBody(mesh, cpt.rigid_body_dofs())

test_matrix = xr.Dataset(coords={
    "period": np.linspace(0.1, 1.0, 300),
    "radiating_dof": list(body.dofs),
    })

for gf in [
        cpt.Delhommeau(tabulation_nz=0, tabulation_nr=0),
        cpt.XieDelhommeau(tabulation_nz=0, tabulation_nr=0),
        # cpt.Delhommeau(),
        # cpt.XieDelhommeau(),
        cpt.Delhommeau(tabulation_nz=200, tabulation_nr=500),
        cpt.XieDelhommeau(tabulation_nz=200, tabulation_nr=500),
        ]:
    solver = cpt.BEMSolver(green_function=gf)
    ds = solver.fill_dataset(test_matrix, body)

    ds.added_mass.sel(radiating_dof="Heave", influenced_dof="Heave").plot(x="period", label=str(gf)[:35] + "...")
plt.legend()
plt.title("Heave added mass ($A_{33}$) of a vertical cylinder")
plt.tight_layout()
plt.show()
```

</details>

**Comment by mancellin** [comment]
The periods we consider above are too high for the resolution of the mesh (several wavelength per panel).

Taking another view on this result:
![Added mass for several Green function implementations](https://github.com/capytaine/capytaine/assets/31126826/96250dfd-c71f-4aab-9cb9-0714b8c53e6f)

The results at wavelength lower than the resolution of the mesh are meaningless. For some reason, the meaningless results from `Delhommeau` are smoother than the meaningless of `XieDelhommeau`, but it is hardly the sign of a bug in `XieDelhommeau`.

**Comment by mancellin** [comment]
I tried to refine the mesh and the issue remains, so my comment just above is probably wrong and there is more to it than just the mesh resolution.

![Added mass for several Green function](https://github.com/capytaine/capytaine/assets/31126826/c050d695-53a5-4ee7-a491-eaf85119e724)

**Comment by mancellin** [comment]
This effect occurs for the heaving cylinder, but not for the surging cylinder. So the issue might be related to horizontal panels.

A candidate cause for the issue is the singularity in the vertical derivative of the wave part of this version of the Green function, which is integrated only with a 1-point scheme.

$$
\frac{\partial \mathcal{G}^+}{\partial \tilde z}(\tilde r, \tilde z) = \mathcal{G}^+(\tilde r, \tilde z) + \frac{2}{\sqrt{\tilde r^2 + \tilde z^2}}
$$

**Comment by mancellin** [comment]
Interesting update with the new thinner tabulation (#439 and #476):

![Figure_1](https://github.com/capytaine/capytaine/assets/31126826/c8007378-726b-49eb-bd0d-04cf0e618f89)

**Comment by mancellin** [comment]
Interesting consequence of #508:

The high-freq variant (formerly Delhommeau) lost its good behavior at high-frequency, but without gaining a better accuracy for panels tangent to the free surface.

The low-freq variant (formerly XieDelhommeau) stays basically the same.

The results below were obtained with 0fa40a85912a12e2cc87654fad91413d5f979c6a

## Test case from this thread

![High frequency cylinder](https://github.com/capytaine/capytaine/assets/31126826/f13b2307-b7cf-4cbe-bfd6-623e4fc55097)


<details><summary>Details</summary>
<p>

```python
import numpy as np
import xarray as xr
import capytaine as cpt
import matplotlib.pyplot as plt

mesh = cpt.mesh_vertical_cylinder(center=(0, 0, 0), length=2.0, radius=1.0).immersed_part()
body = cpt.FloatingBody(mesh, cpt.rigid_body_dofs())

res = cpt.BEMSolver().solve(cpt.RadiationProblem(body=body, period=0, radiating_dof="Heave"))
asymp_added_mass = res.added_mass["Heave"]

test_matrix = xr.Dataset(coords={
    "period": np.linspace(0.1, 1.0, 100),
    "radiating_dof": list(body.dofs),
    })

for gf in [
        cpt.Delhommeau(tabulation_nr=0, gf_singularities="high_freq"),
        cpt.Delhommeau(tabulation_nr=0, gf_singularities="low_freq"),
        # cpt.Delhommeau(tabulation_nr=0, gf_singularities="better_low_freq"),
        cpt.Delhommeau(gf_singularities="high_freq"),
        cpt.Delhommeau(gf_singularities="low_freq"),
        # cpt.Delhommeau(gf_singularities="better_low_freq"),
           ]:

    solver = cpt.BEMSolver(green_function=gf)
    ds = solver.fill_dataset(test_matrix, body)

    ds.added_mass.sel(radiating_dof="Heave", influenced_dof="Heave").plot(x="period", label=gf)

plt.hlines(asymp_added_mass, test_matrix.period.values.min(), test_matrix.period.values.max(), linestyle="--", color="grey", label="Asymptotic")
plt.legend()
plt.title("Heave added mass ($A_{33}$) of a vertical cylinder")
plt.tight_layout()
plt.show()

```

</p>
</details> 

## Test case from #180

![Tangent sphere](https://github.com/capytaine/capytaine/assets/31126826/5a020715-35e0-4733-a9a5-80d996458c5b)

<details><summary>Details</summary>
<p>

```python
import numpy as np
import xarray as xr
import matplotlib.pyplot as plt
import capytaine as cpt

coarse_body = cpt.FloatingBody(mesh=cpt.mesh_sphere(radius=1.0, center=(0, 0, -1.01), resolution=(10, 10)), name="coarse sphere")
coarse_body.add_translation_dof(name="Heave")

finer_body = cpt.FloatingBody(mesh=cpt.mesh_sphere(radius=1.0, center=(0, 0, -1.01), resolution=(30, 30)), name="finer sphere")
finer_body.add_translation_dof(name="Heave")

test_matrix = xr.Dataset(coords={
    'omega': np.linspace(1.0, 3.0, 30),
    'radiating_dof': ["Heave"]
})

for body, linestyle in [
        (coarse_body, ":"),
        (finer_body, "--"),
       ]:
    for gf, color in [
           (cpt.Delhommeau(gf_singularities="high_freq"), "C0"),
           (cpt.Delhommeau(gf_singularities="low_freq"), "C1"),
           ]:
        ds = cpt.BEMSolver(green_function=gf).fill_dataset(test_matrix, body)

        ds['added_mass'].plot(x='omega', color=color, linestyle=linestyle,  label=f"{body.mesh.nb_faces} panels | {gf}")

plt.title("Sphere almost tangent to the free surface")
plt.legend()
plt.tight_layout()
plt.show()
```

</p>
</details>

**Comment by mancellin** [comment]
Same as above but with a purely "high_freq" variant.
The even numbers of points are much more accurate, but the asymptotics looks more reasonable for the odd numbers.

![Added mass of cylinder for seveal variants of the Green function](https://github.com/capytaine/capytaine/assets/31126826/3569ad24-f46a-462c-97fc-359b8bea24d7)

**Comment by mancellin** [comment]
The divergence of the no-tabulation version was due to the even number of points in the Simpson rule.
For some reason, the low_freq coefficients are more sensitive to this than the high_freq ones.

This sensitivity is also visible in the tabulated results, as the interpolation is causing higher issues with low_freq than the high_freq coefficients.
Instead of always building the tabulation for the low_freq version and correct afterwards, when could always build the tabulation for the high_freq version and correct afterwards.
It would make sense that the high frequency behavior would then be improved, but it might cause issues somewhere else?

**Comment by mancellin** [comment]
Trying to summarize what happened here:

- There are two variants of the infinite depth Green function, which are called in Capytaine the `low_freq` variant and the `high_freq` variant (they used to be called by the less descriptive names `XieDelhommeau` and `Dehommeau`).

- As can be expected from their name, the `low_freq` variant is more accurate for frequencies near $\omega=0$, while the `high_freq` variant is more accurate for frequencies near $\omega=\infty$.

- The `high_freq` variant is the default in Nemoh and in Capytaine before version 2.2. The `low_freq` variant is the default in several other codes including WAMIT and HAMS.

- The error of using the wrong variant is amplified by the tabulation of the Green function: in other words, the tabulating the `low_freq` variant make it worse at high frequencies, while it barely affects its low frequency behavior (and vice-versa for the `high_freq` variant). The precise coupling between the tabulation fineness and the error is not obvious.

![Plot comparing the variants at low frequency](https://github.com/capytaine/capytaine/assets/31126826/b6bd9926-feb6-4f50-9571-60a32448ed09)
![Plot comparing the variants at high frequency](https://github.com/capytaine/capytaine/assets/31126826/7444e91b-1df2-482c-9075-1391fce81820)

- There used to be a bug in Capytaine that made the untabulated version worse that it should be. It has been fixed in #515.

- The `low_freq` variant is also more accurate near the free surface.

- The `low_freq` variant is the new default in Capytaine 2.2.

- In the future, we could imagine to automatically switch between the `low_freq` and the `high_freq` variant depending on the frequency to always have the lowest error.
